basic interface / blueprint for the command line handling

diveshuttam opened the pull request:
Refer https://github.com/EasyGnuPG/pgpg/issues/21

This basically **declares** the functions which will handle various commands. (The invocation of the respective functions on a command is done by `click` library so that is also done with this PR)

- The naming of the functions is somewhat inspired from `egpg`'s source.
- There are a few commands in `egpg` like `import` and `pass` which are keywords in python. Interestingly these also have alternate names in `egpg` like `add` and `passphrase` so I have used alternates instead for these commands so that we don't run into any problem later. 
- The alternate names of commands will be added as and when I implement them.
- Currently I am raising a `NotImplementedError` from each of these.
- These function declarations are a part of the command line interface, these will use the library interface of pgpg for performing various tasks. (The library interface will in turn use GPGME bindings)
- I will probably define library interface as I proceed with the project. Since, I already had what all commands we need (through `egpg`). I felt it would be easy to define a skeleton for cli.
- As I am using click library this generates the help command on its own through doc strings. So we have inherently achieved the help command as highlighted below.
- Currently I have not replaced `egpg` with `pgpg` in most of the docstrings we will do as we implement these commands.

```bash
(python3venv) divesh@MY-HP:~/Desktop/gsoc$ pgpg --help
Usage: pgpg [OPTIONS] COMMAND [ARGS]...

  There  are  scads  of options presented by GnuPG, which are all part of
  making it the flexible and powerful an encryption framework that it is.
  But it´s extremely complicated to get started with, and that quite rea‐
  sonably puts people off.

  pgpg is a wrapper script that tries to simplify the  process  of  using
  GnuPG. In order to simplify things, it is opinionated about the "right"
  way to use GnuPG.

Options:
  --help  Show this message and exit.

Commands:
  contact  Commands for handling the contacts.
  gpg      Run any gpg command (but using the...
  info     Display info about the current configuration...
  init     Initialize egpg.
  key      Commands for handling the key.
  migrate  Get keys and contacts from another gpg...
  open     Decrypt and verify the signature of the given...
  seal     Sign and encrypt a file.
  set      Change the settings.
  sign     Sign a file.
  verify   Verify the signature.
```

@vnev and others can start contributing (if they wish) after this is accepted.
Since I have laid most of the framework as I wished it to be. One can start by implementing the functions for the respective command. `click` takes care for calling these functions.
One can find more about click at http://click.pocoo.org/6/, it will be used extensively for handling various checks and aspects of command line. Hence we will be defining option's, arguments etc. using `click`

diveshuttam committed:
.pyc __pycache__ and egginfo added to gitignore

diveshuttam committed:
basic pip setup added

diveshuttam committed:
basic install instructions added

diveshuttam committed:
Merge branch 'master' of github.com:EasyGnuPG/pgpg

update local branch

diveshuttam committed:
add vscode to gitignore

diveshuttam committed:
interface to core cli commands

diveshuttam committed:
key cli

diveshuttam committed:
contact cli

diveshuttam committed:
update readme

diveshuttam committed:
combininig cli

dashohoxha commented:
@diveshuttam can you give me some instructions on how I can test this branch? Please use only **pgpg-ds** for development, and I expect only instructions that use it. The same development instructions should be on `README.md`.

If I run in my computer your code that is still in development and not yet tested, this is not safe. All kinds of bugs can happen that may harm my computer, my home directory, my gpg keys, etc. This is not simply hypothetical, it has happened to me. There was a silly bug in my code that destroyed my computer (something like running `rm -rf /$dir`, where `$dir` was empty due to a bug). What is worse, it destroyed also the machines of my friends that were trying to help me by testing my application.

Since then I have always tried to work on a chroot, and a bit later docker was invented and I moved to docker.

This is why I insist that you (and I and everyone else) develop only inside a container, where it is isolated from the host machine.

diveshuttam committed:
Modified installation instructions

diveshuttam commented:
>@diveshuttam can you give me some instructions on how I can test this branch? Please use only pgpg-ds for development, and I expect only instructions that use it. The same development instructions should be on README.md.

Have updated the readme. You just have to use the instructions in pgpg-ds and you are ready to use this. Please inform if you face any problems.

>If I run in my computer your code that is still in development and not yet tested, this is not safe. All kinds of bugs can happen that may harm my computer, my home directory, my gpg keys, etc. This is not simply hypothetical, it has happened to me. There was a silly bug in my code that destroyed my computer (something like running rm -rf /$dir, where $dir was empty due to a bug). What is worse, it destroyed also the machines of my friends that were trying to help me by testing my application.

Thanks for this, I didn't think of it this way and mostly do testing on my machine only. Currently the code is harmless as it doesn't do anything except printing `--help` but yes I get the potential risk. Will always do the development in the container only.

>I have not checked your code yet, but have you given some thought to extensibility? Because egpg is extensible, and pgpg should be extensible too.

Though I haven't added them to this PR and plan to do this after second eval only, but I did give it an initial thought. 

>Actually, you can use the same mechanism used by egpg for extending it by defining new commands, if you can load bash scripts inside a python program, and then run the bash function that is defined inside this bash script.
Instead of bash script I thought of using python files instead. The structure would be parallel to what egpg code does. Using `.py` would be easier and better as user can use the functions of our module inside `.py` file. If one is comfortable with bash instead, He/She can add small code in `.py` files (pgpg loads only the `.py` files) to execute the required bash script. 
Will add the instructions to this when I add this feature. The probable caveat with using bash scripts is that the user cannot use the python functions inside pgpg (at least no so easily). He will have to accomplish his task through the exposed commands only.

diveshuttam commented:
@dashohoxha @sourabhtk37 I mostly want this PR to be reviewed for code structure and Please give your suggestions for improvement . Though you can test the `--help` command but it is only a side effect of using `click` and organizing the command structure.

sourabhtk37 commented on pgpg/cli.py:
```diff
@@ -1,2 +1,42 @@
+from pgpg.core.cli import cmd_sign
+from pgpg.core.cli import cmd_verify
+from pgpg.core.cli import cmd_set
+from pgpg.core.cli import cmd_gpg
```
convert to single line import. 
`from pgpg.core.cli import (<insert functions here>)`

sourabhtk37 reviewed (commented):
Nothing much to review here.

dashohoxha commented:
> Though I haven't added them to this PR and plan to do this after second eval only, but I did give it an initial thought.

This is an important design decision that may affect the way that you structure your application. So you should not leave it for the end, you should decide about it before starting to code.

> Instead of bash script I thought of using python files instead.

I would prefer to use both of them (bash extensions and python extensions), if possible.

By the way, how can we call a Python function from a Bash script?
If this is possible and not very difficult, then I would actually prefer to keep the code of the application as it is (bash code), and replace only the calls to the command `gpg` with the respective calls to the library functions of GPGME.

> I mostly want this PR to be reviewed for code structure and Please give your suggestions for improvement.

The problem is that I am not sure that your approach is the right one. For example why do we have to use the **click** library? Is it just automatic help and automatic argument parsing that we benefit? For those benefits we have to accept the restrictions that this library imposes (due to the assumptions that it makes).

What about using just plain Python, and the library **getopt**:
 - https://stackoverflow.com/questions/6382804/how-to-use-getopt-optarg-in-python-how-to-shift-arguments-if-too-many-arguments
 - https://github.com/dougsland/python-by-examples/blob/master/getOpt.py
 - https://www.journaldev.com/17508/python-getopt
 - https://docs.python.org/3/library/getopt.html

The advantage of this library is that the same one is used on the existing code of `egpg`, so the translation from Bash code to Python code can be straightforward.

I would also prefer just a translation of the existing code to Python, instead of a complete restructuring and a complete rewrite, if possible.

diveshuttam commented:
>This is an important design decision that may affect the way that you structure your application. So you 
should not leave it for the end, you should decide about it before starting to code.

Ok, I will think about this first, though I have quite a bit of Idea how we can structure this. I think this can be done after checking if pgpg is initialized and `pgpg init` when I have the `.egpg` / `.pggp` folder from which we can load files(scripts). The script files will have similar naming convention as `egpg`. 

> I would prefer to use both of them (bash extensions and python extensions), if possible.

Python extensions are definitely easy, I can directly import the python source and attach it to the concerned command. Bash extensions would also be not difficult for the coding part but they would be less powerful than the python ones I believe though will surely add them.

> By the way, how can we call a Python function from a Bash script?
> If this is possible and not very difficult, then I would actually prefer to keep the code of the application as it is (bash code), and replace only the calls to the command `gpg` with the respective calls to the library functions of GPGME.

I haven't thought much about this, a faint idea is that I can expose them through a python script which can be used to call functions using `exec` ,`eval` calls in python. But definitely it is cumbersome and not very safe, You cannot pass any parameters except text and my script will have to take care of each and every function you want to call and then pass you the things in pre decided format as text back. And all will become messy.

>The problem is that I am not sure that your approach is the right one. For example why do we have to use the **click** library? Is it just automatic help and automatic argument parsing that we benefit? For those benefits we have to accept the restrictions that this library imposes (due to the assumptions that it makes).

I researched about many libraries which can be used for designing command line interfaces in python while working with [termicoder](https://github.com/termicoder/termicoder) as I didn't knew python and was learning back then I did try most of them once.
Though if you wish I can use any of the libraries but **click** gives most of the features built right off, to quote a few:
- delivery of arguments to the right functions. 
  - though this is basic but click takes care to call the right function on a particular command and displaying invalid error if its not there.
- validation of options 
  - I can specify the type of argument that is intended and it will do the checking e.g. if the argument is a path, whether the path exists
- Generation of help
  - Though this is a simple feature it is helpful as I can arbitrarily add and remove commands in future and I don't have to worry about help updation logic.
  - Click also generates help for individual options and arguments
  - All these help written with the function itself helps in the documentation of the code also.

These are just some features, individually they are small but in adding up I expect to get a nice interface and with significant less effort. Regarding restrictions, yes we might have some, but click is heavily customizable through decorators. Though I believe to achieve the functionality required for `pgpg` we don't need any of it. 

>What about using just plain Python, and the library **getopt**:
> - https://stackoverflow.com/questions/6382804/how-to-use-getopt-optarg-in-python-how-to-shift-arguments-if-too-many-arguments
> - https://github.com/dougsland/python-by-examples/blob/master/getOpt.py
> - https://www.journaldev.com/17508/python-getopt
> - https://docs.python.org/3/library/getopt.html

>The advantage of this library is that the same one is used on the existing code of `egpg`, so the translation from Bash code to Python code can be straightforward.

>I would also prefer just a translation of the existing code to Python, instead of a complete restructuring and a complete rewrite, if possible.

I agree that translation may be easier but I would prefer an hybrid approach as Python is much powerful than bash so why not utilize its power. Though in terms of structure I will mostly keep aligned with `egpg` but I feel it's good if I can reduce code on my end. The less code the less errors. Though I am relying on click but it is stable, highly tested and maintained. Same goes with using any other library.
`Getopt` will still require me significant amount of parsing and validation logic in code like calling the right functions on a command, then even after I call right functions they will have to parse their own options like you do in egpg whereas with click it parses and passes them as correct function parameters. 
Though in bash it makes sense as it doesn't have explicit parameter passing but in python but it is good for the functions to have things which it uses present in it declaration(signature) and not to parse them.
This is small but important concern the required functions are a unit they only do what is intended and parsing logic is abstracted out with click. If I avoid this I'll probably end up creating quite a bit of the click's functionality indirectly in each of the function for parsing. Or a common function for parsing and then passing the right parameters. 
On the other hand if I use click and face some restriction, I can add code to circumvent this too but it would be way less than to build the parsing code myself. (or to add to each function)

Also rather than just translation I would prefer to combine best of both your code and structuring as well as what I know. If we face any restrictions we can definitely circumvent these and in the process I will learn something new.

diveshuttam committed:
Combine mulitple imports into one

Brackets have been used refering
https://www.python.org/dev/peps/pep-0328/
Also remove not implemented error.

dashohoxha commented:
Let's try first this task: https://github.com/EasyGnuPG/egpg/issues/49
I think this achieves what we want (replacing the usage of the command `gpg` with GPGME library calls) without changing anything else.

> The time that is left can be use for improving this project.

What do you mean by improving? If you mean adding new features etc. this is not in the scope of this project.

dashohoxha reviewed (approved):
I am merging this PR, so that you can keep experimenting with it. However I am going to move to the branch 'click', so that you can experiment with the other approach as well.

diveshuttam commented:
>Let's try first this task: EasyGnuPG/egpg#49
>I think this achieves what we want (replacing the usage of the command gpg with GPGME library calls) without changing anything else.

Fine lets try this approach also.

>What do you mean by improving? If you mean adding new features etc. this is not in the scope of this project.

Packaging, documenting the library calls, testing this project. As you said in one conversation, exposing this as a library is important which allows people to extend this much more easily.
